src/datetime.ts

```diff
@@ -0,0 +1,161 @@
+export interface DateTimeLike {
+  readonly year: number;
+  readonly month: number;
+  readonly day: number;
+  readonly hour: number;
+  readonly minute: number;
+  readonly second: number;
+  /**
+   * Whether the date and time is in UTC.
+   */
+  readonly utc: boolean;
+}
+
+/**
+ * Represents a date and time. Either local or UTC.
+ */
+export class DateTime implements DateTimeLike {
+  private readonly numeric: number;
+
+  public get year(): number {
+    return Math.floor(this.numeric / 100000000000);
+  }
+
+  public get month(): number {
+    return Math.floor((this.numeric / 1000000000) % 100);
+  }
+
+  public get day(): number {
+    return Math.floor((this.numeric / 10000000) % 100);
+  }
+
+  public get hour(): number {
+    return Math.floor((this.numeric / 100000) % 100);
+  }
+
+  public get minute(): number {
+    return Math.floor((this.numeric / 1000) % 100);
+  }
+
+  public get second(): number {
+    return Math.floor((this.numeric / 10) % 100);
+  }
+
+  public get utc(): boolean {
+    return this.numeric % 10 == 1;
+  }
+
+  private constructor(numeric: number) {
+    this.numeric = numeric;
+  }
+
+  /**
+   * Converts DateTime into Date object:
+   * - If DateTime is UTC, returns a Date object created from UTC time.
+   * - If DateTime is local, returns a Date object with local time.
+   */
+  public toDate(): Date {
+    if (this.utc) {
+      return new Date(
+        Date.UTC(
+          this.year,
+          this.month - 1,
+          this.day,
+          this.hour,
+          this.minute,
+          this.second,
+        ),
+      );
+    } else {
+      return new Date(
+        this.year,
+        this.month - 1,
+        this.day,
+        this.hour,
+        this.minute,
+        this.second,
+      );
+    }
+  }
+
+  /**
+   * Converts DateTime into a plain object.
+   */
+  public toObject(): DateTimeLike {
+    return {
+      year: this.year,
+      month: this.month,
+      day: this.day,
+      hour: this.hour,
+      minute: this.minute,
+      second: this.second,
+      utc: this.utc,
+    };
+  }
+
+  /**
+   * Returns a new DateTime object representing the current date and time.
+   */
+  public static now(): DateTime {
```

Comment:
Since DateTime() doesn't really make sense without knowing the local timezone of the RRuleSet, should `now()` always construct a UTC date instead (using `getUTC*()`)? This way `DateTime.now()` wouldn't be lying in case the RRuleSet is not using in the system's local timezone. We could also have two functions `local(date?: Date = new Date())` and `utc(date?: Date = new Date())` instead, to make it extra clear what is being created.

Reply:
Actually I created this method just to shorten fallback value in `RRuleset`:

```typescript 
this.dtstart = setOrDtstart?.dtstart ?? DateTime.now();
```

instead of this 

```typescript 
this.dtstart = setOrDtstart?.dtstart ?? DateTime.fromObject({
      year: now.getFullYear(),
      month: now.getMonth() + 1,
      day: now.getDate(),
      hour: now.getHours(),
      minute: now.getMinutes(),
      second: now.getSeconds(),
      utc: false,
});
```

In reality I think this is totally useless method. I don't see any use-case when someone will need to create rrule for current date and time.

But I like your idea to remove `now`. I'm not sure about `local` and `utc, since as for me they sound like alternatives to `create` method. 

What if we just introduce methods `DateTime.fromDate(date: Date)`, `DateTime.fromUTCDate(date: Date)`, `DateTime.toUTCDate` and `DateTime.toDate`?

Reply:
Introduced `DateTime.fromDate(date: Date, options?: {utc: boolean})` method

Reply:
This makes sense. However, after thinking more about this, I concluded that converting between `Date` and `DateTime` is potentially unsafe because it has a lot of pitfalls:
A JS `Date` represents a point in time relative to UTC, but a `DateTime` is a date and time without a reference point (unless it's `utc=true`). So, constructing a local `DateTime` from a `Date` is unsafe when the intention is to pass it to `RRuleSet` with a `tzid` different from the system timezone. Similarly, converting a local `DateTime` (received from an `RRuleSet`) to a `Date` is unsafe when it was returned by an `RRuleSet` with a `tzid` different from the system timezone.

Even your example for constructing a Luxon date might be incorrect:
```javascript
luxon.DateTime.fromJSDate(dateTime.toDate(), {zone: dateTime.utc ? 'utc' : undefined})
```

The correct way to construct this would be this:
```
luxon.DateTime.fromJSDate(dateTime.toDate(), {zone: dateTime.utc ? 'utc' : rruleSet.tzid})
```

That makes me think whether we should only allow constructing a utc `DateTime` from a JS `Date` and converting a `DateTime` to a `Date` can only be done through an instance of `RRuleSet`.

Reply:
> So, constructing a local DateTime from a Date is unsafe when the intention is to pass it to RRuleSet with a tzid different from the system timezone. Similarly, converting a local DateTime (received from an RRuleSet) to a Date is unsafe when it was returned by an RRuleSet with a tzid different from the system timezone.

From my point of view it's expected behavior. When you call `datetime.toDate`, the `Date` object becomes attached to node's timezone. When you pass the same `datetime` to  rruleSet, it becomes attached to specified `tzid`. So `datetime` itself has no timezone (unless it's utc) and should be always considered in context.

But I agree that it can be unsafe, unclear and be misused.

> That makes me think whether we should only allow constructing a utc DateTime from a JS Date and converting a DateTime to a Date can only be done through an instance of RRuleSet.

I'd like to leave such date manipulations to some specified date libraries, which will definitely do this better. In terms of working with rrule simple datetime object with utc/non-utc state looks enough.

I suggest if you don't have any use-cases for `toDate`, `fromDate` methods, we can just remove them for current release. I checked my usages - and I don't need them at all. Probably in your case they are unneeded also.

Reply:
I think if we remove `toDate` and `fromDate` it's worth to exclude `utc` from `DateTimeLike`, to make it compatible with `luxon`:

```
toObject(): DateTimeLike
static fromObject(date: DateTimeLike, options?: {utc?: boolean}): DateTime
```

And for serialization purposes introduce:

```
toString(): string // 20240515T183839Z or 20240515T183839 for non-UTC
static fromString(str: string)
```

Reply:
I have a couple places where I‘d convert back and forth between dates and date-time but since we’re using Luxon I could accomplish all this using fromObject, so this would work for me.